Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -1245,38 +1275,258 @@ RefPtr<CSSStyleValue> Parser::parse_basic_shape_value(TokenStream<ComponentValue
+        auto left = parse_length_percentage(arguments_tokens);
+        if (!left.has_value())
+            left = right;
+
+        arguments_tokens.discard_whitespace();
+        if (arguments_tokens.has_next_token())
             return nullptr;
 
-        argument_tokens.discard_whitespace();
-        if (argument_tokens.has_next_token())
+        transaction.commit();
+        return BasicShapeStyleValue::create(Inset { LengthBox(top.value(), right.value(), bottom.value(), left.value()) });
+    }
+
+    if (function_name.equals_ignoring_ascii_case("xywh"sv)) {
+        // xywh() = xywh( <length-percentage>{2} <length-percentage [0,∞]>{2} [ round <'border-radius'> ]? )
+        // FIXME: Parse the border-radius.
+        auto arguments_tokens = TokenStream { component_value.function().value };
+
+        arguments_tokens.discard_whitespace();
+        auto x = parse_length_percentage(arguments_tokens);
+        if (!x.has_value())
+            return nullptr;
+
+        arguments_tokens.discard_whitespace();
+        auto y = parse_length_percentage(arguments_tokens);
+        if (!y.has_value())
+            return nullptr;
+
+        arguments_tokens.discard_whitespace();
+        auto width = parse_length_percentage(arguments_tokens);
+        if (!width.has_value())
+            return nullptr;
+
+        arguments_tokens.discard_whitespace();
+        auto height = parse_length_percentage(arguments_tokens);
+        if (!height.has_value())
+            return nullptr;
+
+        arguments_tokens.discard_whitespace();
+        if (arguments_tokens.has_next_token())
+            return nullptr;
+
+        // Negative width or height is invalid.
+        if ((width->is_length() && width->length().raw_value() < 0)
+            || (width->is_percentage() && width->percentage().value() < 0)
+            || (height->is_length() && height->length().raw_value() < 0)
+            || (height->is_percentage() && height->percentage().value() < 0))
             return nullptr;
 
-        points.append(Polygon::Point { *x_pos, *y_pos });
+        transaction.commit();
+        return BasicShapeStyleValue::create(Xywh { x.value(), y.value(), width.value(), height.value() });
     }
 
-    transaction.commit();
-    return BasicShapeStyleValue::create(Polygon { FillRule::Nonzero, move(points) });
+    if (function_name.equals_ignoring_ascii_case("rect"sv)) {
+        // rect() = rect( [ <length-percentage> | auto ]{4} [ round <'border-radius'> ]? )
+        // FIXME: Parse the border-radius.
+        auto arguments_tokens = TokenStream { component_value.function().value };
+
+        arguments_tokens.discard_whitespace();
+        auto top = parse_length_percentage(arguments_tokens);
+        if (!top.has_value()) {
+            if (arguments_tokens.consume_a_token().is_ident("auto"sv)) {
+                top = Length::make_auto();
+            } else {
+                return nullptr;
+            }
+        }
```

Comment:
I'd be very tempted to make a little lambda that does this, instead of repeating the same steps 4 times.